File: level_2/rabbit_hole_1/TLE_on_last_4/solution.cpp
#include <bits/stdc++.h>
using namespace std;
// Write any include statements here

int find_cycle(int node, vector<int>& L, vector<bool>& visited) {
    int turtoise = L[node];
    int hare = L[L[node]];
    while (turtoise != hare) {
        if (visited[turtoise]) {
            return -1;
        }
        visited[turtoise] = true;
        turtoise = L[turtoise];
        hare = L[L[hare]];
    }
    return turtoise;
}

void mark_cycle(int start, vector<int>& L, vector<bool>& visited, vector<int>& cycle_size) {
    int node = L[start];
    int size = 1;
    while (node != start) {
        visited[node] = true;
        size++;
        node = L[node];
    }
    cycle_size[start] = size;
    node = L[start];
    while (node != start) {
        cycle_size[node] = size;
        node = L[node];
    }
    return;    
}

int getMaxVisitableWebpages(int N, vector<int> L) {
    // Write your code here
    for (int i = 0; i < N; i++) L[i]--;

    vector<int> cycle_size(N, 1);
    vector<bool> sources(N, true);
    vector<bool> visited(N, false);
    int start;
    for (int node = 0; node < N; node++) {
        sources[L[node]] = false;
        if (visited[node]) {
            continue;
        }
        start = find_cycle(node, L, visited);
        if (start != -1) {
            mark_cycle(start, L, visited, cycle_size);
        }
        visited[node] = true;
    }

    int sol = 0;
    int cur, cur_len;
    for (int i = 0; i < N; i++) sol = max(sol, cycle_size[i]);
    for (int node = 0; node < N; node++) {
        if (!sources[node]) {
            continue;
        }
        cur = node;
        cur_len = 0;
        while (cycle_size[cur] == 1) {
            cur = L[cur];
            cur_len++;
        }
        cur_len += cycle_size[cur];
        sol = max(sol, cur_len);
    }
    return sol;
}

File: level_2/rabbit_hole_1/TLE_on_last_4/solution.py
from typing import List, Optional
# Write any import statements here

def getMaxVisitableWebpages(N: int, L: List[int]) -> int:
    # Write your code here
    def find_cycle(node: int) -> Optional[int]:
        turtoise = L[node]
        hare = L[L[node]]
        while turtoise != hare:
            if visited[turtoise]:
                return
            visited[turtoise] = True
            turtoise = L[turtoise]
            hare = L[L[hare]]
        return turtoise
        
    def mark_cycle(start: int) -> None:
        node = L[start]
        size = 1
        while node != start:
            visited[node] = True
            size += 1
            node = L[node]

        cycle_size[start] = size
        node = L[start]
        while node != start:
            cycle_size[node] = size
            node = L[node]


    L = [x - 1 for x in L]

    cycle_size = [1 for _ in range(N)]
    sources = {i for i in range(N)}
    visited = [False for _ in range(N)]
    for node in range(N):
        if L[node] in sources:
            sources.remove(L[node])
        if visited[node]:
            continue
        start = find_cycle(node)
        if start is not None:
            mark_cycle(start)
        
        visited[node] = True

    sol = max(cycle_size)
    for node in sources:
        cur = node
        cur_len = 0
        while cycle_size[cur] == 1:
            cur = L[cur]
            cur_len += 1
        cur_len += cycle_size[cur]
        sol = max(sol, cur_len)

    return sol


if __name__ == "__main__":
    L = [2,4,2,3,1,1,6]
    L = [4, 1, 2, 1]
    L = [2, 4, 2, 2, 3]
    L = [4, 3, 5, 1, 2]
    print(getMaxVisitableWebpages(len(L), L))